### ISSUE TITLE
AP feature: payload_transparency

### ISSUE THREAD

**Issue by oberstet** [wamp-proto/wamp-proto#228]
This feature is meant for two use cases:

* end-to-end payload encryption
* performance optimization

Normally, WAMP application payload is transmitted in `args` (list) and/or `kwargs` (dict) items in a subset of WAMP messages.

With `payload_transparency`, the application payload is transmitted as a combined (no distinct positional vs keyword arguments) and string/binary valued, opaque item.

E.g. normally, the WAMP `PUBLISH` message can take these forms:

* `[PUBLISH, Request|id, Options|dict, Topic|uri]`
* `[PUBLISH, Request|id, Options|dict, Topic|uri, Arguments|list]`
* `[PUBLISH, Request|id, Options|dict, Topic|uri, Arguments|list, ArgumentsKw|dict]`

When `payload_transparency` is enabled, a fourth variant is possible:

* `[PUBLISH, Request|id, Options|dict, Topic|uri, Payload|string or binary]`

The `Payload` item may be of type string or binary, and is left totally uninterpreted by a router.

> The `Payload` will be of string-type when using the JSON serializer, and must contain a base64 encoded value of the original binary, opaque application payload to be transmitted. With binary-transparent serializers like CBOR MessagePack, the  `Payload` may be of type binary also.

**Comment by oberstet** [comment]
When using `payload_transparency`, certain features a router might have like transparently translating between different serialization formats might not be available anymore.

However, the *use* of `payload_transparency` can happen at a message-per-message basis.

The availability of the feature must be announce by including a `payload_transparency` attribute in the role features dictionary with a `true` value.

**Feature announcement**

A client announcing `payload_transparency` signals to a router that it may want to use this feature in the future. However, it may only do so if the router has announced `payload_transparency` too. When a client uses `payload_transparency` without the router supporting it, the results are undefined (from error to bogus behavior). A client that has not announced `payload_transparency` should not use the feature later, even if the router supports it. A client that has announced `payload_transparency` is in no way required or not even expected to use the feature. The client just signals it's ability to use. A router must not send messages of the fourth form to a client unless it has announced `payload_transparency`.

**Comment by oberstet** [comment]
A less nice aspect of above approach is: we've managed to avoid polymorphic messages up to now (messages that have the same number of items, but differ in type of item). The fourth form however has the same number of items as the existing form that carries only "args" (no "kwargs").

An alternative would be to set `args=null` and `kwargs=null`, and make the fourth form one element **more** than the existing variants with both `args` and `kwargs`.

**Comment by oberstet** [comment]
The following WAMP messages carry application payload (have `args` and `kwargs` variants):

* PubSub: `PUBLISH`, `EVENT`
* RPC: `CALL`, `INVOCATION`, `YIELD`, `RESULT`, `ERROR`

All of these messages become another variant that carris a `payload` item instead of `args` and `kwargs`.

**Comment by oberstet** [comment]
There is an implementation in Autobahn|Python and Crossbar.io now (https://github.com/crossbario/autobahn-python/issues/572 and https://github.com/crossbario/crossbar/issues/570)

**Comment by konsultaner** [comment]
I just implemented this feature into my router/client. it helped me a lot with being compatible to other frameworkes that already have a communication interface and need to translate to WAMP. It is also very helpful for my end-to-end encryption project.

There is one thing I came accross with. The websocket transport channel has a limited/defined message size. It would help a lot if there was a way to also split the payload accross multiple messages if a certain size was reached. At the moment I have some extra logic that solves that issue. It would just be very helpful if the client automatically chunks at a certain size and the receiving client combines the payload automatically without adding some extra busines logic.

+1 for adding it to the specs, awesome feature

**Comment by oberstet** [comment]
> The websocket transport channel has a limited/defined message size

No, WebSocket does not have a limited message size. The size of WebSocket frames is limited .. but the limit is 2^63.

Regarding end-to-end encryption, there is https://github.com/wamp-proto/wamp-proto/issues/229 - an alpha flavor of this is already in AutobahnPython and Crossbar.io.

**Comment by konsultaner** [comment]
@oberstet Thank your for that hint. I use Netty 4.0 as the underlying WebSocketServer/Client which for some [unknown reason](http://stackoverflow.com/questions/37207133/why-is-does-netty-4-0-x-limit-the-default-websocket-frame-size-to-65536-byte) has a default famesize limit of 64kbyte.

**Comment by oberstet** [comment]
A framesize limit doesn't matter anyways, as a WS server needs to assemble all frames for a message. In any case: this is all the task of the transport, and WAMP should not replicate such things. Also, rather than inventing your own end-to-end encryption, I'd rather welcome you in discussing / advancing the WAMP-cryptosign effort;) Well, obvious I guess.

**Comment by oberstet** [comment]
As it happens, two more use cases came up for "payload transparency" (on #autobahn IRC, raised by users, and real world scenarios):

* being able to use custom types with serializers that support that on the wire (namely, CBOR), and in conjunction with implementations of thereof that expose this (eg see [here](https://github.com/agronholm/cbor2#customizing-encodingdecoding))
* with WAMP routers that provide a MQTT bridge, and in scenarios where existing MQTT client must be supported without any modification, the MQTT payload essentially cannot be interpreted/known by the router, and hence there is little point in not transferring the payload "as is" without touching it at all (which obviously is good for performance too)

So these two use cases, together with the original use case of end-to-end encryption, demonstrate the general usefulness.

An essentially, this allows one to choose the tradeoff one is willing to make on a flexible basis:

* either profit from the automatic conversion between different serializers, payload scrubbing/normalization (for added security/robustness as it protects clients), automatic payload validation against schemata and so on, when there is no payload transparency in place
* or, when using payload transparency, cover the mentioned use cases above

The tradeoffs are inherent and unavoidable, but WAMP puts users into power and allows them to choose what's best in a specific situation.
